File: crawl/full_domain_crawler.js
#!/usr/bin/env node
/**
 * Full-domain BFS crawler using Playwright (Chromium via CDP).
 * Crawls every reachable page on each faculty domain.
 * Saves raw text + structured JSON — same pattern as the original CS crawl.
 *
 * Usage: npx playwright install chromium && node full_domain_crawler.js
 */

const { chromium } = require('playwright');
const fs = require('fs');
const path = require('path');
const crypto = require('crypto');
const { URL } = require('url');

const OUT_DIR = path.join(process.env.HOME, 'Documents/Claude/projects/unimelb-current-students-ia/crawl/full-faculty');
const MAX_PAGES = 500;
const MAX_DEPTH = 6;

const FACULTIES = [
  // law already done (266 pages)
  ['feit', 'eng.unimelb.edu.au'],
  ['arts', 'arts.unimelb.edu.au'],
  ['science', 'science.unimelb.edu.au'],
  ['education', 'education.unimelb.edu.au'],
  ['fbe', 'fbe.unimelb.edu.au'],
  ['ffam', 'finearts-music.unimelb.edu.au'],
  ['mdhs', 'mdhs.unimelb.edu.au'],
  ['msd', 'msd.unimelb.edu.au'],
];

fs.mkdirSync(OUT_DIR, { recursive: true });

function fingerprint(url) {
  return crypto.createHash('sha1').update(url).digest('hex').slice(0, 16);
}

function normalizeUrl(url, base) {
  try {
    const u = new URL(url, base);
    u.hash = '';
    return u.href;
  } catch {
    return null;
  }
}

function isInternal(url, domain) {
  try {
    const u = new URL(url);
    return u.hostname === domain || u.hostname.endsWith('.' + domain);
  } catch {
    return false;
  }
}

function shouldSkip(url) {
  // Skip binary files, auth-gated, external, anchors
  if (!url) return true;
  const skipPatterns = [
    /\.(pdf|docx?|xlsx?|pptx?|zip|tar|gz|jpg|jpeg|png|gif|svg|mp4|mp3|webm|css|js|ico|xml|rss)$/i,
    /mailto:/i,
    /^javascript:/i,
    /^#/,
    /tel:/i,
    /my\.unimelb/i,
    /lms\.unimelb/i,
    /canvas\.unimelb/i,
    /login/i,
  ];
  return skipPatterns.some(p => p.test(url));
}

async function crawlPage(page, url) {
  console.log(`  [FETCH] ${url.slice(0, 100)}`);
  
  try {
    await page.goto(url, { waitUntil: 'domcontentloaded', timeout: 20000 });
  } catch (e) {
    console.log(`    NAV FAIL: ${e.message.slice(0, 80)}`);
    return null;
  }
  
  // Wait for any JS rendering
  await page.waitForTimeout(2000);
  
  // Get page title
  let title = '';
  try { title = await page.title(); } catch {}
  
  // Get body text
  let text = '';
  try {
    text = await page.evaluate(() => {
      const main = document.querySelector('main') || document.querySelector('article') || document.body;
      if (!main) return '';
      const clone = main.cloneNode(true);
      clone.querySelectorAll('script, style, nav, footer, header, .uom-megamenu, .site-nav, .breadcrumb, [role="navigation"]').forEach(el => el.remove());
      return (clone.innerText || clone.textContent || '').trim();
    });
  } catch {}
  
  const wordCount = text ? text.split(/\s+/).filter(w => w.length > 0).length : 0;
  
  if (wordCount === 0) {
    console.log(`    EMPTY — likely WAF or render failure`);
    return null;
  }
  
  // Extract internal links
  let links = [];
  try {
    links = await page.evaluate((baseUrl) => {
      const links = new Set();
      document.querySelectorAll('a[href]').forEach(a => {
        try {
          const u = new URL(a.href, baseUrl);
          u.hash = '';
          links.add(u.href);
        } catch {}
      });
      return [...links];
    }, url);
  } catch {}
  
  // Filter links
  const currentDomain = new URL(url).hostname;
  const internalLinks = links
    .filter(l => isInternal(l, currentDomain))
    .filter(l => !shouldSkip(l))
    .map(l => normalizeUrl(l, url))
    .filter(Boolean);
  
  // Deduplicate
  const uniqueLinks = [...new Set(internalLinks)];
  
  // Save raw text
  const fp = fingerprint(url);
  const rawDir = path.join(OUT_DIR, `raw_${currentDomain.split('.')[0]}`);
  fs.mkdirSync(rawDir, { recursive: true });
  fs.writeFileSync(path.join(rawDir, `${fp}.txt`), 
    `URL: ${url}\nTITLE: ${title}\nWORD_COUNT: ${wordCount}\n\n${text}`);
  
  return {
    url,
    normalizedUrl: url,
    title,
    wordCount,
    outboundLinks: uniqueLinks,
    fingerprint: fp,
  };
}

async function crawlDomain(browser, slug, domain) {
  const base = `https://${domain}`;
  const visited = new Set();
  const frontier = [base + '/'];
  const pages = [];
  let pageCount = 0;
  
  console.log(`\n${'='.repeat(60)}`);
  console.log(`CRAWLING: ${slug} (${domain})`);
  console.log(`${'='.repeat(60)}`);
  
  const page = await browser.newPage();
  await page.setViewportSize({ width: 1440, height: 900 });
  
  // Quick check: can we load the homepage?
  try {
    await page.goto(base + '/', { waitUntil: 'domcontentloaded', timeout: 15000 });
    const title = await page.title();
    console.log(`  Homepage loaded: "${title}"`);
  } catch (e) {
    console.log(`  CANNOT LOAD HOMEPAGE: ${e.message.slice(0, 100)}`);
    await page.close();
    return { domain, slug, error: e.message, pages_crawled: 0 };
  }
  
  while (frontier.length > 0 && pageCount < MAX_PAGES) {
    const url = frontier.shift();
    if (visited.has(url)) continue;
    visited.add(url);
    
    const pageData = await crawlPage(page, url);
    
    if (pageData) {
      pages.push(pageData);
      pageCount++;
      
      // Add new links to frontier
      const newLinks = pageData.outboundLinks.filter(l => !visited.has(l) && !frontier.includes(l));
      frontier.push(...newLinks);
      
      console.log(`    [${pageCount}] ${pageData.wordCount} words, +${newLinks.length} links, frontier=${frontier.length}`);
    }
    
    // Polite delay
    await page.waitForTimeout(800);
  }
  
  await page.close();
  
  // Save structured JSON
  const output = {
    domain,
    slug,
    pages_crawled: pages.length,
    urls_visited: visited.size,
    frontier_remaining: frontier.length,
    pages,
  };
  
  const outPath = path.join(OUT_DIR, `${slug}.json`);
  fs.writeFileSync(outPath, JSON.stringify(output, null, 2));
  console.log(`  SAVED: ${pages.length} pages → ${outPath}`);
  
  return output;
}

async function main() {
  console.log('Starting full-domain faculty crawler...');
  console.log(`Output: ${OUT_DIR}\n`);
  
  // Connect to existing Chrome via CDP or launch new browser
  let browser;
  try {
    browser = await chromium.connectOverCDP('http://localhost:9222');
    console.log('Connected to existing Chrome via CDP');
  } catch (e) {
    console.log('No existing Chrome CDP — launching Chromium...');
    browser = await chromium.launch({ headless: true });
  }
  
  const allResults = {};
  
  for (const [slug, domain] of FACULTIES) {
    try {
      allResults[slug] = await crawlDomain(browser, slug, domain);
    } catch (e) {
      console.log(`\n  FATAL for ${slug}: ${e.message}`);
      allResults[slug] = { domain, slug, error: e.message, pages_crawled: 0 };
    }
  }
  
  // Save summary
  const summaryPath = path.join(OUT_DIR, '_summary.json');
  fs.writeFileSync(summaryPath, JSON.stringify(allResults, null, 2));
  
  console.log(`\n\n${'='.repeat(60)}`);
  console.log('FULL CRAWL COMPLETE');
  console.log(`${'='.repeat(60)}`);
  for (const [slug, data] of Object.entries(allResults)) {
    if (data.error) {
      console.log(`  ${slug}: ERROR — ${data.error.slice(0, 60)}`);
    } else {
      console.log(`  ${slug}: ${data.pages_crawled} pages, ${data.urls_visited} visited`);
    }
  }
  console.log(`\nData: ${OUT_DIR}/`);
  
  await browser.close();
}

main().catch(e => { console.error(e); process.exit(1); });

File: crawl/full_domain_crawler.py
#!/usr/bin/env python3
"""
Full-domain BFS crawler via CDP (Chrome DevTools Protocol).
Crawls every reachable page on a domain, extracts text + links,
saves raw corpus and structured JSON.
Pattern: same approach as the original CS crawl, but for the full domain.
"""
import asyncio
import json
import time
import os
import hashlib
import re
import urllib.request
import websockets
from urllib.parse import urljoin, urlparse

CDP_BASE = "http://localhost:9222"
OUT_DIR = os.path.expanduser("~/Documents/Claude/projects/unimelb-current-students-ia/crawl/full-faculty")

# 9 faculties (corrected FEIT domain)
FACULTIES = [
    ("law", "law.unimelb.edu.au"),
    ("feit", "eng.unimelb.edu.au"),
    ("arts", "arts.unimelb.edu.au"),
    ("science", "science.unimelb.edu.au"),
    ("education", "education.unimelb.edu.au"),
    ("fbe", "fbe.unimelb.edu.au"),
    ("ffam", "finearts-music.unimelb.edu.au"),
    ("mdhs", "mdhs.unimelb.edu.au"),
    ("msd", "msd.unimelb.edu.au"),
]

MAX_PAGES_PER_DOMAIN = 500  # Safety cap
PAGE_TIMEOUT = 15

os.makedirs(OUT_DIR, exist_ok=True)

EXTRACT_LINKS_JS = """
(() => {
    const links = [];
    const base = location.origin;
    document.querySelectorAll('a[href]').forEach(a => {
        let href = a.href;
        if (!href) return;
        // Normalize: resolve relative, strip hash
        try {
            const u = new URL(href, base);
            u.hash = '';
            href = u.href;
        } catch(e) { return; }
        // Only same-domain links
        if (href.startsWith(base + '/') || href === base + '/' || href === base) {
            links.push(href);
        }
    });
    return JSON.stringify([...new Set(links)]);
})
"""

EXTRACT_CONTENT_JS = """
(() => {
    // Get main content: prefer main/article, fall back to body
    const main = document.querySelector('main') || document.querySelector('article') || document.body;
    if (!main) return JSON.stringify({title: document.title, text: '', wordCount: 0});
    
    // Remove scripts, styles, nav, footer
    const clone = main.cloneNode(true);
    clone.querySelectorAll('script, style, nav, footer, header, .uom-megamenu, .site-nav, .breadcrumb').forEach(el => el.remove());
    
    const text = clone.innerText || clone.textContent || '';
    const words = text.trim().split(/\\s+/).filter(w => w.length > 0);
    
    return JSON.stringify({
        title: document.title || '',
        url: location.href,
        text: text.trim().slice(0, 50000),
        wordCount: words.length,
    });
})
"""

def page_fingerprint(url):
    return hashlib.sha1(url.encode()).hexdigest()[:16]

class CDPCrawler:
    def __init__(self, ws_url, domain, slug):
        self.ws_url = ws_url
        self.domain = domain
        self.slug = slug
        self.base = f"https://{domain}"
        self.visited = set()
        self.frontier = [f"{self.base}/"]
        self.pages = []
        self.raw_dir = os.path.join(OUT_DIR, f"raw_{slug}")
        os.makedirs(self.raw_dir, exist_ok=True)
        self.ws = None
        
    async def connect(self):
        self.ws = await websockets.connect(self.ws_url, max_size=20*1024*1024,
            ping_interval=None, close_timeout=10)
        await self.send_recv({"id": 1, "method": "Page.enable"})
    
    async def send_recv(self, msg, expect_id=None):
        await self.ws.send(json.dumps(msg))
        # For commands with an id, wait for the response
        # For events-only (no id), return immediately
        cmd_id = msg.get("id")
        
    async def navigate_and_wait(self, url):
        """Navigate and wait for DOM content loaded."""
        await self.ws.send(json.dumps({"id": 100, "method": "Page.navigate", "params": {"url": url}}))
        
        # Drain messages until domContentEventFired or timeout
        for _ in range(50):
            raw = await asyncio.wait_for(self.ws.recv(), timeout=PAGE_TIMEOUT)
            data = json.loads(raw)
            method = data.get("method", "")
            if method == "Page.domContentEventFired":
                return True
            elif method == "Page.loadEventFired":
                pass  # Keep waiting for domContentEventFired
            # else: ignore other events
        return False
    
    async def eval_js(self, expression):
        """Evaluate JS and return the result value."""
        await self.ws.send(json.dumps({
            "id": 200, "method": "Runtime.evaluate",
            "params": {"expression": expression, "returnByValue": True}
        }))
        raw = await asyncio.wait_for(self.ws.recv(), timeout=10)
        data = json.loads(raw)
        result = data.get("result", {}).get("result", {})
        if result.get("type") == "string":
            return result["value"]
        return None
    
    async def crawl_page(self, url):
        """Crawl a single page: navigate, extract content + links, save raw."""
        fp = page_fingerprint(url)
        print(f"  [{len(self.pages)+1}] {url[:100]}")
        
        # Navigate
        ok = await self.navigate_and_wait(url)
        if not ok:
            print(f"    WARN: navigation timeout")
        
        # Wait for JS rendering
        await asyncio.sleep(2)
        
        # Extract content
        content_raw = await self.eval_js(EXTRACT_CONTENT_JS)
        content = json.loads(content_raw) if content_raw else {}
        
        if not content or content.get("wordCount", 0) == 0:
            print(f"    EMPTY page (wordCount=0) — likely WAF block or JS render failure")
            # Save raw HTML as fallback
            try:
                html = await self.eval_js("document.documentElement.outerHTML.slice(0, 10000)")
                if html:
                    raw_path = os.path.join(self.raw_dir, f"{fp}.html")
                    with open(raw_path, "w") as f:
                        f.write(html)
            except:
                pass
            return None, []
        
        # Extract links
        links_raw = await self.eval_js(EXTRACT_LINKS_JS)
        links = json.loads(links_raw) if links_raw else []
        links = [l for l in links if l not in self.visited]  # Filter visited
        
        # Save raw text
        raw_path = os.path.join(self.raw_dir, f"{fp}.txt")
        with open(raw_path, "w") as f:
            f.write(f"URL: {url}\nTITLE: {content.get('title','')}\nWORD_COUNT: {content.get('wordCount',0)}\n\n")
            f.write(content.get("text", ""))
        
        # Build page record
        page = {
            "url": url,
            "normalizedUrl": url,
            "title": content.get("title", ""),
            "wordCount": content.get("wordCount", 0),
            "outboundLinks": links,
            "fingerprint": fp,
        }
        
        return page, links
    
    async def crawl(self):
        """Main BFS crawl loop."""
        print(f"\n{'='*60}")
        print(f"CRAWLING: {self.slug} ({self.domain})")
        print(f"Frontier start: {len(self.frontier)} URLs")
        print(f"{'='*60}")
        
        await self.connect()
        
        page_count = 0
        while self.frontier and page_count < MAX_PAGES_PER_DOMAIN:
            url = self.frontier.pop(0)
            if url in self.visited:
                continue
            
            self.visited.add(url)
            
            try:
                page, links = await self.crawl_page(url)
            except Exception as e:
                print(f"    ERROR: {e}")
                continue
            
            if page:
                self.pages.append(page)
                page_count += 1
                
                # Add new links to frontier
                new_links = [l for l in links if l not in self.visited and l not in self.frontier]
                self.frontier.extend(new_links)
            
            # Small delay between pages
            await asyncio.sleep(1)
        
        # Save structured data
        out_path = os.path.join(OUT_DIR, f"{self.slug}.json")
        output = {
            "domain": self.domain,
            "slug": self.slug,
            "pages_crawled": len(self.pages),
            "urls_visited": len(self.visited),
            "frontier_remaining": len(self.frontier),
            "pages": self.pages,
        }
        with open(out_path, "w") as f:
            json.dump(output, f, indent=2)
        
        print(f"\n  DONE: {len(self.pages)} pages saved to {out_path}")
        print(f"  Visited: {len(self.visited)}, Frontier remaining: {len(self.frontier)}")
        return output
    
    async def close(self):
        if self.ws:
            await self.ws.close()


async def main():
    # Create a fresh Chrome tab
    req = urllib.request.Request(f"{CDP_BASE}/json/new?url=about:blank", method="PUT")
    with urllib.request.urlopen(req) as resp:
        tab = json.loads(resp.read())
    ws_url = tab["webSocketDebuggerUrl"]
    print(f"Tab created: {tab['id']}")
    
    all_results = {}
    
    for slug, domain in FACULTIES:
        crawler = CDPCrawler(ws_url, domain, slug)
        try:
            result = await crawler.crawl()
            all_results[slug] = result
        except Exception as e:
            print(f"\n  FATAL for {slug}: {e}")
            all_results[slug] = {"error": str(e), "domain": domain}
        finally:
            await crawler.close()
        
        # Create a new tab for each faculty
        req = urllib.request.Request(f"{CDP_BASE}/json/new?url=about:blank", method="PUT")
        with urllib.request.urlopen(req) as resp:
            tab = json.loads(resp.read())
        ws_url = tab["webSocketDebuggerUrl"]
        print(f"\nNew tab: {tab['id']}")
    
    # Save master summary
    summary_path = os.path.join(OUT_DIR, "_summary.json")
    with open(summary_path, "w") as f:
        json.dump(all_results, f, indent=2, default=str)
    
    print(f"\n\n{'='*60}")
    print(f"FULL CRAWL COMPLETE")
    print(f"{'='*60}")
    for slug, data in all_results.items():
        if "error" in data:
            print(f"  {slug}: ERROR - {data['error']}")
        else:
            print(f"  {slug}: {data['pages_crawled']} pages, {data['urls_visited']} visited")
    print(f"\nData saved to: {OUT_DIR}/")

if __name__ == "__main__":
    asyncio.run(main())
